core/consensus/pow/src/lib.rs

```diff
@@ -331,27 +359,35 @@ pub fn start_mine<B: BlockT<Hash=H256>, C, Algorithm, E>(
 					e
 				),
 			}
-
 			std::thread::sleep(std::time::Duration::new(1, 0));
 		}
 	});
 }
 
-fn mine_loop<B: BlockT<Hash=H256>, C, Algorithm, E>(
+fn mine_loop<B: BlockT<Hash=H256>, C, Algorithm, E, SO>(
 	block_import: &mut BoxBlockImport<B>,
 	client: &C,
 	algorithm: &Algorithm,
 	env: &mut E,
 	preruntime: Option<&Vec<u8>>,
 	round: u32,
+	sync_oracle: &mut SO,
+	build_time: std::time::Duration,
 	inherent_data_providers: &inherents::InherentDataProviders,
 ) -> Result<(), String> where
 	C: HeaderBackend<B> + AuxStore,
 	Algorithm: PowAlgorithm<B>,
 	E: Environment<B>,
 	E::Error: std::fmt::Debug,
+	SO: SyncOracle,
 {
 	'outer: loop {
+		if sync_oracle.is_major_syncing() {
+			debug!(target: "pow", "Skipping proposal due to sync.");
+			std::thread::sleep(std::time::Duration::new(1, 0));
+			continue 'outer
+		}
+
 		let best_hash = client.info().best_hash;
 		let best_header = client.header(BlockId::Hash(best_hash))
```

Comment:
We should be using a `SelectChain` here with `LongestChain`. There are some edge cases where the best block as reported by the client may be a little outdated.